crates/uv-auth/src/keyring.rs

```diff
@@ -33,77 +38,160 @@ pub enum Error {
     ParseFailed(#[from] std::string::FromUtf8Error),
 }
 
-/// Get credentials from keyring for given url
-///
-/// See `pip`'s KeyringCLIProvider
-/// <https://github.com/pypa/pip/blob/ae5fff36b0aad6e5e0037884927eaa29163c0611/src/pip/_internal/network/auth.py#L102>
-pub fn get_keyring_subprocess_auth(url: &Url) -> Result<Option<Credential>, Error> {
-    let host = url.host_str();
-    if host.is_none() {
-        return Err(Error::NotKeyringTarget(
-            "Should only use keyring for urls with host".to_string(),
-        ));
+impl KeyringProvider {
+    /// Fetch credentials for the given [`Url`] from the keyring.
+    ///
+    /// Returns `None` if no password was found, even if a username is present on the URL.
+    pub(crate) fn fetch(&self, url: &Url) -> Result<Option<Credentials>, Error> {
+        let host = url.host_str();
+        if host.is_none() {
+            return Err(Error::NotKeyringTarget(
+                "Should only use keyring for urls with host".to_string(),
+            ));
+        }
+
+        if url.password().is_some() {
+            return Err(Error::NotKeyringTarget(
+                "Url already contains password - keyring not required".to_string(),
+            ));
+        }
+
+        debug!("Checking keyring for credentials for `{}`", url.to_string(),);
+        let password = match self {
+            Self::Disabled => Ok(None),
+            Self::Subprocess => self.fetch_subprocess(url),
+            #[cfg(test)]
+            Self::Dummy(provider) => self.fetch_dummy(provider, url),
+        }?;
+
+        Ok(password.map(|password| Credentials::new(url.username().to_string(), Some(password))))
     }
-    if url.password().is_some() {
-        return Err(Error::NotKeyringTarget(
-            "Url already contains password - keyring not required".to_string(),
-        ));
+
+    /// Fetch from the `keyring` subprocess.
+    ///
+    /// See pip's implementation
+    /// <https://github.com/pypa/pip/blob/ae5fff36b0aad6e5e0037884927eaa29163c0611/src/pip/_internal/network/auth.py#L102>
+    fn fetch_subprocess(&self, url: &Url) -> Result<Option<String>, Error> {
```

Comment:
```suggestion
    #[instrument]
    fn fetch_subprocess(&self, url: &Url) -> Result<Option<String>, Error> {
```

To track where we're losing 80ms

```
$ hyperfine -i "keyring get https://pypi.org konstin"
Benchmark 1: keyring get https://pypi.org konstin
  Time (mean ± σ):      80.8 ms ±   5.7 ms    [User: 66.2 ms, System: 11.7 ms]
  Range (min … max):    73.8 ms …  96.2 ms    30 runs
 
  Warning: Ignoring non-zero exit code.
```